riscv/decode_macros.h

```diff
@@ -321,3 +321,35 @@ inline long double to_f(float128_t f) { long double r; memcpy(&r, &f, sizeof(r))
   reg_t h##field = get_field(STATE.henvcfg->read(), HENVCFG_##field)
 
 #endif
+
+#define software_check(x, tval) (unlikely(!(x)) ? throw trap_software_check_fault(tval) : (void) 0)
+#define ZICFILP_READ_XLPE(v, prv) \
+  ({ \
+    reg_t csr_val, mask; \
+    if (p->extension_enabled('S')) { \
+      if (v) { \
+        switch (prv) { \
```

Comment:
See `DECLARE_XENVCFG_VARS` in this file. I suspect it could simplify a lot of this.

Reply:
Sorry, I can't see how the approach taken by `DECLARE_XENVCFG_VARS` could help in an obvious way. In this `xLPE` case, it is essentially a runtime table lookup from an integer pair (virtual, privilege) to a (CSR id, CSR bit mask) pair, which is quite different from the paradigm of `DECLARE_XENVCFG_VARS`.

If you have a good idea about this part, please do let me know =)

Reply:
See my [two commits here](https://github.com/scottj97/riscv-isa-sim/commits/pr-1582-my-xenvcfg/)

Reply:
Fixed.